Build system enhancements

shindere opened the pull request:
The initial intent behind this PR was to start preparing the transition to
autoconf. More specifically, in autoconf-based packages, a few build
variables (e.g. CFLAGS, CPPFLAGS, LDFLAGS) are expected to be
available to the "user" (the person
compiling the package).

So the initial aim of this PR was to renaim the above mentionned build
variables so that they become private to the build system and so that
they can later be re-introduced but without any value so that users can
freely use them without
unintentionnally breaking the build system.

So this PR does that, plus a few additiolnal things in the build system of
the runtime:

- For the debugging, intstrumentation, profiling and PIC related flags,
  clarify which are compiler and which are preprocessor flags

- Factorise the rules to compile C files. This is to make it easier
  to re-introduce the user variables later by decreasing the number of
  places in the build system where they need to be mentionned

@Adrien-N: you may be interested in reviewing this

dra27 commented on asmrun/Makefile:
```diff
@@ -142,20 +142,26 @@ install:
-%.p.$(O): %.c
-	$(CC) -c $(PFLAGS) $(CPPFLAGS) $(OUTPUTOBJ)$@ $<
+$(foreach object_type, $(object_types), \
+  $(eval $(call COMPILE_C_FILE,$(object_type))))
```
I know it's equivalent, but part of me would prefer the definition of the build rule (via `eval`) to be physically below the setting up of dependencies (in this case the env. variables)

damiendoligez replied:
I agree with @dra27: for readability it should be moved down.

dra27 commented on asmrun/Makefile:
```diff
@@ -191,13 +197,13 @@ depend:
+	  sed -e 's/\.o/.p.$$(O)/' >> .depend
+	$(CC) -MM $(OC_CPPFLAGS) $(OC_DEBUG_CPPFLAGS) $^ | \
+	  sed -e 's/\.o/.d.$$(O)/' >> .depend
+	$(CC) -MM $(OC_CPPFLAGS) $(OC_INSTR_CPPFLAGS) $^ | \
```
In all three of these, I think it may be relevant to have had `PFLAGS`, `DFLAGS` and `IFLAGS` first - doesn't GCC go with the first option encountered? In other words, switching it around I think causes CPPFLAGS always to win?

(I'm not convinced this matters in practice, I'm just wondering if the change is necessary)

dra27 commented on byterun/Makefile:
```diff
@@ -42,7 +42,7 @@ endif
 
 ifdef BOOTSTRAPPING_FLEXLINK
-CFLAGS += -DBOOTSTRAPPING_FLEXLINK
+OC_CPPFLAGS += -DBOOTSTRAPPING_FLEXLINK
```
Oops, as the original author of that error 😊

dra27 commented on byterun/Makefile:
```diff
@@ -207,28 +206,39 @@ libcamlrun_pic.$(A): $(PICOBJS)
+
+%.i.$(O): OC_CPPFLAGS += $(OC_INSTR_CPPFLAGS)
+
+%.pic.$(O): OC_CFLAGS += $(SHAREDLIB_CFLAGS)
```
It feels like this should be shared between `asmrun` and `byterun`, rather than duplicated? Given that we're getting into the realm of GNU make macros!

dra27 commented on byterun/Makefile:
```diff
@@ -207,28 +206,39 @@ libcamlrun_pic.$(A): $(PICOBJS)
+	$(CC) -MM $(IFLAGS) $(OC_CPPFLAGS) *.c | sed -e 's/\.o/.i.$$(O)/'>>.$@
+	$(CC) -MM $(OC_CPPFLAGS) $(OC_INSTR_CPPFLAGS) *.c | \
+	  sed -e 's/\.o/.i.$$(O)/' >> .$@
+	$(CC) -MM $(OC_CPPFLAGS) *.c | sed -e 's/\.o/.pic.$$(O)/'>>.$@
```
Same comment about ordering as in `asmrun/Makefile`

dra27 reviewed (commented):
Unless 4.08 is *definitely* going to have the `configure` script available for Windows, I'm dubious of changes which involve renaming variables in `config/Makefile` as it increases still further the headache for configuring Windows-builds (as demonstrated by the changes needed to the Inria CI scripts here!).

Sorry about the comment on the shed colour, but as it's its introduction: I found the `OC_` prefix a little odd to read - can't we just have `OCAML_`?

For now, I wonder if this scheme might be a more compatible first step:
 - Don't rename `SHAREDCCCOMPOPTS` (but I do agree that `SHAREDLIB_CFLAGS` is a much better name, once that name is truly internal to the build system!)
 - Keep `CFLAGS`, `CPPFLAGS` and `LDFLAGS` in `config/Makefile` (and the 4 Windows `Makefile`s, therefore)
 - At the top of `Makefile`s which use `CFLAGS`, `CPPFLAGS`, and `LDFLAGS`, add `OC_CFLAGS=$(CFLAGS)` ... with the intention that those lines get removed when `config/Makefile` starts generating the "correct internal files.

A question on the `CFLAGS` stuff - I thought that the autoconf idea was that the user should be able to specify additional flags which they'd like here (or override them). So in the places where `$(CFLAGS)` is being *changed* to `$(OC_FLAGS)` shouldn't it be becoming `$(CFLAGS) $(OC_FLAGS)` - i.e. we start to move the flags we determined during `configure`, and which are recorded in `config/Makefile` to `OC_FLAGS` but we still allow the user to specify `CFLAGS` themselves? Note that this idea is orthogonal to my third suggestion above (adding `OC_FLAGS=$(CFLAGS)`) as that would duplicate the flags given in `CFLAGS`, obviously.

shindere commented:
@dra27 many thanks for your prompt and detailed review!

David Allsopp (2018/06/15 02:23 -0700):
> Unless 4.08 is *definitely* going to have the `configure` script
> available for Windows, I'm dubious of changes which involve renaming
> variables in `config/Makefile` as it increases still further the
> headache for configuring Windows-builds (as demonstrated by the
> changes needed to the Inria CI scripts here!).

Sure, I understand your point. To be honest, I *definitely* want to
reach this, but at the same time do not dare to promise anything -- one
never knows what can happen.

> Sorry about the comment on the shed colour, but as it's its
> introduction: I found the `OC_` prefix a little odd to read - can't we
> just have `OCAML_`?

Thanks for asking, because I'm really not sure. Let me explain how I
came to choose this.

As you certainly remember, we already have OCAMMLC_CFLAGS etc. and
OCAMLOPT_CFLAGS etc. So, I thought that something like OCAML_CFLAGS
might (wrongly) convey the idea that it's somehow top-level related,
which was one reason why I was not so found of it.

My other thought was that since OC_ is shorter than OCAML_ it would have
a smaller impact on the layout of the code.

And finally, I liked this precisely because it's odd, as you point out.
What I mean is that, to me, such a name really sounds internal and would
discourage me to use it, was I outside of the package, which is
precisely my intention.

What do you think? Are you convinced? :)

> For now, I wonder if this scheme might be a more compatible first step:
>  - Don't rename `SHAREDCCCOMPOPTS` (but I do agree that
>  - `SHAREDLIB_CFLAGS` is a much better name, once that name is truly
>  - internal to the build system!)

What do you mean by "truly internal to the build system"? When wil it
become truly internal, actually?

>  - Keep `CFLAGS`, `CPPFLAGS` and `LDFLAGS` in `config/Makefile` (and the 4 Windows `Makefile`s, therefore)
>  - At the top of `Makefile`s which use `CFLAGS`, `CPPFLAGS`, and
>  - `LDFLAGS`, add `OC_CFLAGS=$(CFLAGS)` ... with the intention that
>  - those lines get removed when `config/Makefile` starts generating
>  - the "correct internal files.

If autoconf comes soon, at least before 4.08, do you agree we wouldn't
need these?

> A question on the `CFLAGS` stuff - I thought that the autoconf idea
> was that the user should be able to specify additional flags which
> they'd like here (or override them). So in the places where
> `$(CFLAGS)` is being *changed* to `$(OC_FLAGS)` shouldn't it be
> becoming `$(CFLAGS) $(OC_FLAGS)` - i.e. we start to move the flags we
> determined during `configure`, and which are recorded in
> `config/Makefile` to `OC_FLAGS` but we still allow the user to specify
> `CFLAGS` themselves? Note that this idea is orthogonal to my third
> suggestion above (adding `OC_FLAGS=$(CFLAGS)`) as that would duplicate
> the flags given in `CFLAGS`, obviously.

You are correct, this PR does not make use of CFLAGS etc., because I
wanted to do a bit more build system refactoring before re-introducing
the variables.

Regarding the order, it should be `$(OC_CFLAGS) $(CFLAGS)` rather than
`$(CFLAGS) $(OC_CFLAGS)` because, at least for gcc it's the later option
that takes precedence, not the first one.

> -%.p.$(O): %.c
> -	$(CC) -c $(PFLAGS) $(CPPFLAGS) $(OUTPUTOBJ)$@ $<
> +$(foreach object_type, $(object_types), \
> +  $(eval $(call COMPILE_C_FILE,$(object_type))))
> 
> I know it's equivalent, but part of me would prefer the definition of
> the build rule (via `eval`) to be physically below the setting up of
> dependencies (in this case the env. variables)

I'd say: if there is something else that needs to be changed in this PR
before it gets merged, then I can change that too. If, after our
discussion, it is considered that this PR can be merged as is, then I'd
rather not change just that, which doesn't mean it can't be changed
later.

> > @@ -191,13 +197,13 @@ depend:
>  	$(error Dependencies cannot be regenerated using the MSVC ports)
>  else
>  depend: $(COBJS:.$(O)=.c) $(sources)
> -	$(CC) -MM $(CFLAGS) $(CPPFLAGS) $^ | sed -e 's/\.o/.$$(O)/' > .depend
> -	$(CC) -MM $(PFLAGS) $(CPPFLAGS) $^ | sed -e 's/\.o/.p.$$(O)/' \
> -	  >> .depend
> -	$(CC) -MM $(DFLAGS) $(CPPFLAGS) $^ | sed -e 's/\.o/.d.$$(O)/' \
> -	  >> .depend
> -	$(CC) -MM $(IFLAGS) $(CPPFLAGS) $^ | sed -e 's/\.o/.i.$$(O)/' \
> -	  >> .depend
> +	$(CC) -MM $(OC_CFLAGS) $(OC_CPPFLAGS) $^ | sed -e 's/\.o/.$$(O)/' > .depend
> +	$(CC) -MM $(OC_CPPFLAGS) $(OC_PROF_CPPFLAGS) $^ | \
> +	  sed -e 's/\.o/.p.$$(O)/' >> .depend
> +	$(CC) -MM $(OC_CPPFLAGS) $(OC_DEBUG_CPPFLAGS) $^ | \
> +	  sed -e 's/\.o/.d.$$(O)/' >> .depend
> +	$(CC) -MM $(OC_CPPFLAGS) $(OC_INSTR_CPPFLAGS) $^ | \
> 
> In all three of these, I think it may be relevant to have had
> `PFLAGS`, `DFLAGS` and `IFLAGS` first - doesn't GCC go with the first
> option encountered?

No, the later one takes precedence.

> In other words, switching it around I think causes
> CPPFLAGS always to win?

Well CFLAGS and CPPFLAGS are intended to contain distinct sets of flags
so their order shouldn't matter.

> -%.$(DBGO): %.c
> -	$(CC) -c $(DFLAGS) $(CPPFLAGS) $(OUTPUTOBJ)$@ $<
> +object_types := %.$(O) %.d.$(O) %.i.$(O) %.pic.$(O)
> +
> +$(foreach object_type, $(object_types), \
> +  $(eval $(call COMPILE_C_FILE,$(object_type))))
> +
> +%.$(DBGO): OC_CPPFLAGS += $(OC_DEBUG_CPPFLAGS)
> +
> +%.i.$(O): OC_CPPFLAGS += $(OC_INSTR_CPPFLAGS)
> +
> +%.pic.$(O): OC_CFLAGS += $(SHAREDLIB_CFLAGS)
> 
> It feels like this should be shared between `asmrun` and `byterun`,
> rather than duplicated? Given that we're getting into the realm of GNU
> make macros!

It's intended to be shared, ultimately, yes. It's just that I didn't
wnat to make this PR even bigger than it already is.

>  .PHONY: depend
>  ifeq "$(TOOLCHAIN)" "msvc"
>  depend:
>  	$(error Dependencies cannot be regenerated using the MSVC ports)
>  else
>  depend: prims.c caml/opnames.h caml/jumptbl.h caml/version.h
> -	$(CC) -MM $(CFLAGS) $(CPPFLAGS) *.c | sed -e 's/\.o/.$$(O)/'>.$@
> -	$(CC) -MM $(DFLAGS) $(CPPFLAGS) *.c | sed -e 's/\.o/.d.$$(O)/'>>.$@
> -	$(CC) -MM $(IFLAGS) $(CPPFLAGS) *.c | sed -e 's/\.o/.i.$$(O)/'>>.$@
> -	$(CC) -MM $(PICFLAGS) $(CPPFLAGS) *.c | sed -e 's/\.o/.pic.$$(O)/'>>.$@
> +	$(CC) -MM $(OC_CPPFLAGS) *.c | sed -e 's/\.o/.$$(O)/'>.$@
> +	$(CC) -MM $(OC_CPPFLAGS) $(OC_DEBUG_CPPFLAGS) *.c | \
> +	  sed -e 's/\.o/.d.$$(O)/'>>.$@
> +	$(CC) -MM $(IFLAGS) $(OC_CPPFLAGS) *.c | sed -e 's/\.o/.i.$$(O)/'>>.$@
> +	$(CC) -MM $(OC_CPPFLAGS) $(OC_INSTR_CPPFLAGS) *.c | \
> +	  sed -e 's/\.o/.i.$$(O)/' >> .$@
> +	$(CC) -MM $(OC_CPPFLAGS) *.c | sed -e 's/\.o/.pic.$$(O)/'>>.$@
> 
> Same comment about ordering as in `asmrun/Makefile`

Same response :) The later flags take precedence. :)

Thanks again for having gone through all this so promptly and with such
good eyes! Warmly appreciated!

shindere commented:
@dra27: regarding your initial remark about the changes to
`config/Makefile`. Given really expect the migrationto autoconf will be
complete before the 4.08 freeze, I'd like to make the following proposal.

Assuming the state of `config/Makefile` is less important between the
releases, could we go with the PR as it is given that, in the unlikely case
where the migration to autoconf would not be completed, then I would do a
commit to restore `config/Makefile`in the state it was before?

Does that look as a reasonable compromise to you?

shindere commented:
Rebased on latest trunk and made sure this PR does not introduce any
check-typo violation, now that our trunk is clean ;-)

xavierleroy commented on asmrun/Makefile:
```diff
@@ -142,20 +142,26 @@ install:
+$(1): %.c
+	$$(CC) -c $$(OC_CFLAGS) $$(OC_CPPFLAGS) $$(OUTPUTOBJ)$$@ $$<
+endef
 
```
I find this GNU make macro wizardry very hard to read.  The result of the macro expansion (i.e. 5 implicit rules) would be more readable, not to mention more compact...

xavierleroy reviewed (commented):
I'm globally OK with the renaming of make variables.  I'd like to be sure all the variables are documented somewhere (name and intended meaning).  There is some effort in this direction in `config/Makefile-templ` but it's probably incomplete and outdated.

shindere commented:
Xavier Leroy (2018/06/18 06:53 -0700):
> xavierleroy commented on this pull request.
> 
> I'm globally OK with the renaming of make variables.  I'd like to be
> sure all the variables are documented somewhere (name and intended
> meaning).  There is some effort in this direction in
> `config/Makefile-templ` but it's probably incomplete and outdated.

Yeah I think the proper place to document all these will be the future
Makefile.config.in where we can add comments for each variable. Roughly
speaking, this file will be an updated version of
`config/Makefile-templ`.

Can the doc wait until this is in place?

> > @@ -142,20 +142,26 @@ install:
>  $(LINKEDFILES): %.c: ../byterun/%.c
>  	$(LN) $< $@
>  
> -%.d.$(O): %.c
> -	$(CC) -c $(DFLAGS) $(CPPFLAGS) $(OUTPUTOBJ)$@ $<
> +# The COMPILE_C_FILE function below receives as argument the pattern
> +# that corresponds to the name of the generated object file
> +define COMPILE_C_FILE
> +$(1): %.c
> +	$$(CC) -c $$(OC_CFLAGS) $$(OC_CPPFLAGS) $$(OUTPUTOBJ)$$@ $$<
> +endef
>  
> 
> I find this GNU make macro wizardry very hard to read.  The result of
> the macro expansion (i.e. 5 implicit rules) would be more readable,
> not to mention more compact...

I have to disagree here. :)

Ultimately this macro will be used to generate all the C files compiling
rules in both asmrun and byterun.

Moreover, this macro is going to become more complex (with the
re-introduction of CFLAGS and CPPFLAGS) so expanding it does really
sound a bad idea to me as far as maintainability is concerned.

So, I'd like to politely insist on the need to keep the macro. :)


Many thanks for having taken the time to review this PR by the way!

damiendoligez commented:
>Can the doc wait until this is in place?

I'd rather not wait, for two reasons:
* documentation makes reviewing easier
* you'll be really sad at the end of the project when you have to write all the docs at once

shindere commented:
Damien Doligez (2018/06/19 08:20 -0700):
> >Can the doc wait until this is in place?
> 
> I'd rather not wait, for two reasons:
> * documentation makes reviewing easier
> * you'll be really sad at the end of the project when you have to
> * write all the docs at once

My suggestion was not to wait until after everything is done. My
suggestion was to include the doc in the autoconf PR itslef rather than
now. Woudl that be okay?

damiendoligez commented on asmrun/Makefile:
```diff
@@ -191,13 +197,14 @@ depend:
-	  >> .depend
-	$(CC) -MM $(IFLAGS) $(CPPFLAGS) $^ | sed -e 's/\.o/.i.$$(O)/' \
-	  >> .depend
+	$(CC) -MM $(OC_CFLAGS) $(OC_CPPFLAGS) $^ | \
```
Didn't you forget to remove `$(OC_CFLAGS)`?

damiendoligez commented on manual/tools/Makefile:
```diff
@@ -1,8 +1,7 @@
-COMPFLAGS=-I $(OTOPDIR)/otherlibs/str -I $(OTOPDIR)/otherlibs/unix
-include $(TOPDIR)/Makefile.tools
+uTOPDIR=../..  COMPFLAGS=-I $(OTOPDIR)/otherlibs/str -I
+$(OTOPDIR)/otherlibs/unix include $(TOPDIR)/Makefile.tools
```
Fat fingers? Your text editor has re-wrapped these three lines into two and you have an extraneous `u` at the beginning of the first one. This probably explains the Appveyor failure that we saw earlier today.

damiendoligez reviewed (changes requested):
Looks good except for a few things you should look into.

shindere committed:
Improve the handling of debugging flags for runtime build

shindere committed:
Improve the handling of profiling flags for runtime build

shindere committed:
Improve the handling of PIC flags for runtime build

shindere committed:
Improve the handling of instrumentation flags for runtime build

shindere committed:
Rename C compiler related build variables

This commit renames a few C compiler related build variables so that
they are reserved for the build system. They will then be re-introduced,
but this time as user varialbes whose value can be freely customized
when compiling the package, without risking to conflict with those
command-line flags that are required by the build system itself.

Here are the variables this commit renames:

- CFLAGS -> OC_CFLAGS
- CPPFLAGS -> OC_CPPFLAGS
- LDFLAGS -> OC_LDFLAGS

Note: before this commit the compilation of scheduler.c in
otherlibs/threads was relying on make's implicit rule to compile C files.

Since this commit stops using the standard variables for flags,
it is necessary to introduce an explicit rule to compile C files
and that makes use of the newly introduced variables.

shindere committed:
byterun/Makefile: add preprocessor flags to the right varialbes

A few -D and -I flags were added to OC_CFLAGS while they should be added
to OC_CPPFLAGS

As a consequence, there is no need to pass OC_CFLAGS to the compiler
when computing dependencies for C files.

shindere committed:
asmrun/Makefile: don't use OC_CFLAGS when computing libasmrun dependencies

shindere committed:
Rename build variable SHAREDCCCOMPOPTS to SHAREDLIB_CFLAGS

shindere committed:
{asmrun,byterun}/Makefile: merge rules compiling C files

shindere committed:
Add Changes entry for GPR#1840 (this one)

shindere commented:
@damiendoligez thanks a lot for the careful review and all my apologies
for the remaining errors.

Damien Doligez (2018/06/20 02:28 -0700):
> damiendoligez requested changes on this pull request.
> 
> Looks good except for a few things you should look into.
> 
> 
> >  
> -%.p.$(O): %.c
> -	$(CC) -c $(PFLAGS) $(CPPFLAGS) $(OUTPUTOBJ)$@ $<
> +$(foreach object_type, $(object_types), \
> +  $(eval $(call COMPILE_C_FILE,$(object_type))))
> 
> I agree with @dra27: for readability it should be moved down.

Done in both `asmrun/Makefile` and `byterun/Makefile`.
 
> > @@ -191,13 +197,14 @@ depend:
>  	$(error Dependencies cannot be regenerated using the MSVC ports)
>  else
>  depend: $(COBJS:.$(O)=.c) $(sources)
> -	$(CC) -MM $(CFLAGS) $(CPPFLAGS) $^ | sed -e 's/\.o/.$$(O)/' > .depend
> -	$(CC) -MM $(PFLAGS) $(CPPFLAGS) $^ | sed -e 's/\.o/.p.$$(O)/' \
> -	  >> .depend
> -	$(CC) -MM $(DFLAGS) $(CPPFLAGS) $^ | sed -e 's/\.o/.d.$$(O)/' \
> -	  >> .depend
> -	$(CC) -MM $(IFLAGS) $(CPPFLAGS) $^ | sed -e 's/\.o/.i.$$(O)/' \
> -	  >> .depend
> +	$(CC) -MM $(OC_CFLAGS) $(OC_CPPFLAGS) $^ | \
> 
> Didn't you forget to remove `$(OC_CFLAGS)`?

Indeed, many thanks for your vigilance. This is now fixed.

> > @@ -1,8 +1,7 @@
> -TOPDIR=../..
> -COMPFLAGS=-I $(OTOPDIR)/otherlibs/str -I $(OTOPDIR)/otherlibs/unix
> -include $(TOPDIR)/Makefile.tools
> +uTOPDIR=../..  COMPFLAGS=-I $(OTOPDIR)/otherlibs/str -I
> +$(OTOPDIR)/otherlibs/unix include $(TOPDIR)/Makefile.tools
> 
> Fat fingers? Your text editor has re-wrapped these three lines into
> two and you have an extraneous `u` at the beginning of the first one.
> This probably explains the Appveyor failure that we saw earlier today.

That has now been fixed too, thanks for reporting.

What remains is to understand why this has not been caught by the other
CI solutions, apparently. Any hint?

shindere commented:
Damien Doligez (2018/06/20 06:59 -0700):
> damiendoligez approved this pull request.

Thanks! Merged.
